File: Python-server/src/controllers/playlistSongsController.py
import os
from flask import request, jsonify
from sqlalchemy.orm import Session
from werkzeug.utils import secure_filename
from config.db import get_db
from models.playlist import Playlist
from models.playlistSongs import PlaylistSongs
from models.song import Song
from boto3 import client
from dotenv import load_dotenv
from io import BytesIO
from datetime import datetime

load_dotenv()

def addPlaylistSong():
    try:
        # Obtener los datos del formulario
        id_playlist = request.form.get('id_playlist')
        id_song = request.form.get('id_song')

        if not id_playlist or not id_song:
            return jsonify({"message": "All fields are required", "status": False}), 400

        db: Session = next(get_db())

        # Verificar si la playlist existe
        existing_playlist = db.query(Playlist).filter_by(id=id_playlist).first()
        if not existing_playlist:
            return jsonify({"message": "Playlist not found", "status": False}), 404

        # Verificar si la canción existe
        existing_song = db.query(Song).filter_by(id=id_song).first()
        if not existing_song:
            return jsonify({"message": "Song not found", "status": False}), 404

        # Verificar si la relación ya existe
        existing_relation = db.query(PlaylistSongs).filter_by(id_playlist=id_playlist, id_song=id_song).first()
        if existing_relation:
            return jsonify({"message": "Song already exists on this playlist", "status": False}), 400

        # Crear la relación en la base de datos
        new_relation = PlaylistSongs(id_playlist=id_playlist, id_song=id_song)
        db.add(new_relation)
        db.commit()

        return jsonify({"message": "Song added successfully", "relation": new_relation.to_dict(), "status": True}), 201

    except Exception as error:
        return jsonify({"message": "Error adding song", "error": str(error), "status": False}), 500

def getSongsByPlaylist(id_playlist):
    try:
        if not id_playlist:
            return jsonify({"message": "Playlist ID is required", "status": False}), 400

        db: Session = next(get_db())

        # Consultar la playlist con las canciones asociadas
        playlist = db.query(Playlist).filter_by(id=id_playlist).first()

        if not playlist:
            return jsonify({"message": "Playlist not found", "status": False}), 404

        # Obtener las canciones asociadas a la playlist
        songs = db.query(Song).join(PlaylistSongs).filter(PlaylistSongs.id_playlist == id_playlist).all()

        if not songs:
            return jsonify({"message": "This playlist doesn't have any songs", "status": False}), 404

        # Retornar las canciones asociadas a la playlist
        songs_dict = [song.to_dict() for song in songs]
        return jsonify({"songs": songs_dict, "status": True}), 200

    except Exception as error:
        return jsonify({"message": "Error fetching songs", "error": str(error), "status": False}), 500

def deletePlaylistSong(id_playlist, id_song):
    try:
        if not id_playlist or not id_song:
            return jsonify({"message": "Playlist ID and Song ID are required", "status": False}), 400

        db: Session = next(get_db())

        # Eliminar la relación entre la playlist y la canción
        deleted = db.query(PlaylistSongs).filter_by(id_playlist=id_playlist, id_song=id_song).delete()

        if not deleted:
            return jsonify({"message": "Song not found in the playlist", "status": False}), 404

        db.commit()

        return jsonify({"message": "Song was successfully removed from the playlist", "status": True}), 200

    except Exception as error:
        return jsonify({"message": "Error deleting song from playlist", "error": str(error), "status": False}), 500

File: Nodejs-server/src/controllers/playlistSongsController.js
const PlayListSongs = require('../models/playlistSongs');
const Song = require('../models/song');
const Playlist = require('../models/playlist');
// Crear una nueva relación entre una canción y una playlist

const addPlaylistSong = async (req, res) => {
    try {
        const { id_playlist, id_song } = req.body;

        if (!id_playlist || !id_song) {
            return res.status(400).json({ message: "All fields are required", status: false });
        }
        const existingPlaylist = await Playlist.findByPk(id_playlist);
        if (!existingPlaylist) {
            return res.status(404).json({ message: "Playlist not found", status: false });
        }
        const existingSong = await Song.findByPk(id_song);
        if (!existingSong) {
            return res.status(404).json({ message: "Song not found", status: false });
        }

        const existingRelation = await PlayListSongs.findOne({ where: { id_playlist, id_song } });
        if (existingRelation) {
            return res.status(400).json({ message: "Song already exists on this playlist", status: false });
        }

        const relation = await PlayListSongs.create({
            id_playlist,
            id_song
        });
        res.status(201).json({ message: "Song added successfully", relation, status: true });
    } catch (error) {
        res.status(500).json({ message: "Error adding song", error, status: false });
    }
}

// Obtener todas las canciones de la playlist de un usuario
const getSongsByPlaylist = async (req, res) => {
    try {
        const { id_playlist } = req.params;

        if (!id_playlist) {
            return res.status(400).json({ message: "Playlist ID is required", status: false });
        }

        const playlist = await Playlist.findByPk(id_playlist, {
            include: {
                model: Song,
                through: { attributes: [] } // Excluye las columnas de la tabla de unión
            }
        });

        if (!playlist) {
            return res.status(404).json({ message: "Playlist not found", status: false });
        }

        if (playlist.Songs.length === 0) {
            return res.status(404).json({ message: "This playlist doesn't have any songs", status: false });
        }

        // Retornar las canciones asociadas a la playlist
        res.status(200).json(playlist.Songs);
    } catch (error) {
        res.status(500).json({ message: "Error fetching songs", error, status: false });
    }
}

// Eliminar una canción de la playlist
const deletePlaylistSong = async (req, res) => {
    try {
        const { id_playlist, id_song } = req.params;

        if (!id_playlist || !id_song) {
            return res.status(400).json({ message: "Playlist ID and Song ID are required", status: false });
        }

        const deleted = await  PlayListSongs.destroy({
            where: { id_playlist: id_playlist, id_song: id_song }
        });

        if (!deleted) {
            return res.status(404).json({ message: "Song not found in the playlist", status: false });
        }

        res.status(200).json({ message: "Song was successfully removed from the playlist", status: true });
    } catch (error) {
        res.status(500).json({ message: "Error deleting song from playlist", error, status: false });
    }
};
module.exports = {
    addPlaylistSong,
    getSongsByPlaylist,
    deletePlaylistSong
};